### PR TITLE
tutorial: added binding tutorial

### PR DESCRIPTION
**Pull request by Piankero**
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [x] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [x] Documentation is concise and good to read
- [x] Examples are well chosen and understandable

### CODE REVIEW

**Comment by Piankero** [comment]
@markus2330 I added this branch to your repository since other poeple will also work on it

@raphi011 @PhilippGackstatter you can work on the binding tutorial here

File: doc/dev/language-binding.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+error code itself since the hierarchy is integrated in it. For example you can check if the code starts with
+`C01...` to catch all `Permanent Errors`.
+
+### Error Message
+
+In Elektra every error has a predefined format. You can take a look at the [related design decision](../decisions/error_message_format.md)
+to see how it looks like. In case of an error at least the following part has to be returned:
+
+```
+Sorry, module `MODULE` issued [error|warning] `NR`:
```
Comment (markus2330) [inline doc/dev/language-binding.md]: It is unclear if it should be written with or without backticks. Define first the accessors and then write down how the whole message looks like using the accessors.

**Comment by Piankero** [comment]
> Please also write some introduction for how to write plugin bindings (as you did in Java)

I have not written the java binding and also do not know how to do it. I just patched it to use the error codes

**Comment by markus2330** [comment]
@Piankero Please do the renames and rebase on master so that others can continue.

**Comment by Piankero** [comment]
> Please do the renames and rebase on master so that others can continue.

I already did the renaming if i am not mistaken

**Comment by markus2330** [comment]
@PhilippGackstatter can you add some text now?

**Comment by PhilippGackstatter** [comment]
> @PhilippGackstatter can you add some text now?

Sure! I will get started as soon as possible for the things I have experience in. I can't yet talk about plugin bindings or for the tools, but can add them later.

I can definitely provide a guide for the CMake integration.

For part 3, these bullet points differ in implementation for every language, so what exactly should I talk about here? Come up with a complete list of things that should be provided? I guess I'll get started and then we'll see.

**Comment by markus2330** [comment]
> Sure! I will get started as soon as possible for the things I have experience in.

Thank you!

> I can definitely provide a guide for the CMake integration.

That would be wonderful as it is sometimes one of the biggest challenges.

> For part 3, these bullet points differ in implementation for every language, so what exactly should I talk about here? Come up with a complete list of things that should be provided? I guess I'll get started and then we'll see.

It should be as language-neutral as possible as obviously nobody will implement the same Rust binding again. You could talk about the design decisions you made for type safety as this might be relevant for other languages.

Btw. you can also write a tutorial about how to **use** the Rust binding but this should not be part of this tutorial. In this tutorial there should be nothing of interest for a user of your binding. Instead it should be focused on people writing other bindings.

File: doc/tutorials/language-bindings.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+If you have a language which does not support inheritance this way like GoLang, you can still use the
+error code itself since the hierarchy is integrated in it. For example you can check if the code starts with
+`C01...` to catch all `Permanent Errors`.
+
+### Error Message
+
+In Elektra every error has a predefined format. You can take a look at the [related design decision](../decisions/error_message_format.md)
+to see how it looks like.
+
+Every Exception/Error struct/etc. should have separate accessors to individual parts of the message.
```
Comment (markus2330) [inline doc/tutorials/language-bindings.md]: @Piankero There is no information about how to implement the getters. (That the error information is in the meta-keys of the parentKey).

**Comment by markus2330** [comment]
@kodebach can you here also write about how to write bindings for the highlevel API? It would be great if you also write about the goals of the high-level API (that elektraGet never fails) because for some languages it might make more sense to reimplement what is done in the highlevel lib (like it was done in the C++ API).

**Comment by PhilippGackstatter** [comment]
I'm trying to push a commit, but I'm getting

```
ERROR: Permission to ElektraInitiative/libelektra.git denied to PhilippGackstatter.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

@markus2330 
Can you grant me the rights for this branch?

**Comment by markus2330** [comment]
I do not think this is possible. We usually do not create branches here (@Piankero forgot about this).

Simply pull it to your repo and make a PR from there. We will remove this branch after we have merged your PR.

**Comment by kodebach** [comment]
> how to write bindings for the highlevel API?

I will add some information. Although it should be pretty straight forward.

> for some languages it might make more sense to reimplement what is done in the highlevel lib

Unless the language uses a type system completely different from the one used by the highlevel API, reimplementing the API doesn't really make sense. It would just create a lot of work to keep everything in sync.

On the other hand it might actually make more sense to create completely new APIs with different goals for other languages. Many modern languages have quite good error handling concepts, so the need for an API call that cannot fail isn't really there. For example in Rust the equivalent of `elektraGetLong` could easily return a `Result<i32, Error>`, since the `?`  operator makes handling the `Result` very straightforward and still very safe.

**Comment by markus2330** [comment]
> I will add some information.

Thank you!

> Unless the language uses a type system completely different from the one used by the highlevel API, reimplementing the API doesn't really make sense.

This might be the most important information. It would also be interesting if the code generator `kdb gen` should be used or another generator preferred instead.

**Comment by kodebach** [comment]
For the code-generated part of the high-level API, we will most likely need new templates for `kdb gen` for other languages. (Also a reason, why I made the code-generator somewhat flexible).

For example for C++ we could generate nested `structs` with overridden operators.

**Comment by markus2330** [comment]
Yes, it would be nice to port the current functionality of pythongen to `kdb gen`. But the first step will be generation of docu (man pages, html pages, ...) out of specifications. Jakob will work on this.

**Pull request closed without merging**